components/fs/fat/fs_diskio.c

```diff
@@ -0,0 +1,290 @@
+	return STA_NOINIT;
+}
+
+DSTATUS disk_status (
+	BYTE pdrv		/* Physical drive nmuber to identify the drive */
+)
+{
+	DSTATUS stat;
+	int result;
+
+	switch (pdrv) {
+	default:
+		return RES_OK;
+	}
+	return STA_NOINIT;
+}
+
+DRESULT disk_ioctl (BYTE pdrv, BYTE cmd, void* buff) {
+	DRESULT res;
+	switch (pdrv) {
+	default:
+		if (cmd == GET_SECTOR_SIZE) {
+			WORD *size = buff;
+			*size = config->sector_size;
+			res = RES_OK;
+		}
+		if (cmd == CTRL_SYNC) {
+			res = RES_OK;
+			#ifdef FS_DEBUG_PRINT
+			sddf_printf("blk_enqueue_syncreq\n");
+			#endif
+			blk_enqueue_req(blk_queue_handle, BLK_REQ_FLUSH, 0, 0, 0, co_get_handle());
+			blk_request_pushed = true;
+			co_block();
+			res = (DRESULT)(uintptr_t)co_get_args();
+		}
+	}
+	return res;
+}
+
+#ifdef MEMBUF_STRICT_ALIGN_TO_BLK_TRANSFER_SIZE
+
+#define MOD_POWER_OF_2(a, b) ((a) & ((b) - 1))
+#define DIV_POWER_OF_2(a, b) ((a) >> (__builtin_ctz(b)))
+#define MUL_POWER_OF_2(a, b) ((a) << (__builtin_ctz(b)))
+
+DRESULT disk_read(BYTE pdrv, BYTE *buff, LBA_t sector, UINT count) {
+	DRESULT res;
+	switch (pdrv) {
+		default: {
+			int handle = co_get_handle();
+			// Accroding the protocol, all the read/write addr passed to the blk_virt should be page aligned
+			// Substract the handle with one as the work coroutine ID starts at 1, not 0
+			uint64_t read_data_offset = coroutine_blk_addr[handle - 1];
+			uint16_t sector_size = config->sector_size;
+			// This is the same as BLK_TRANSFER_SIZE / sector_size
+			uint16_t sector_per_transfer = DIV_POWER_OF_2(BLK_TRANSFER_SIZE, sector_size);
+			uint32_t sddf_sector = DIV_POWER_OF_2(sector, sector_per_transfer);
+			// The final sddf_count is always positive, however in the process of calculating it may be added with a negative value, so use signed integer here
+			int32_t sddf_count = 0;
+			uint32_t unaligned_head_sector = MOD_POWER_OF_2(sector_per_transfer - MOD_POWER_OF_2(sector, sector_per_transfer), sector_per_transfer);
+			uint32_t unaligned_tail_sector = MOD_POWER_OF_2(sector + count, sector_per_transfer);
+			if (unaligned_head_sector) {
+				sddf_count += 1;
+			}
+			if (unaligned_tail_sector) {
+				sddf_count += 1;
+			}
+			// DIV_POWER_OF_2(aligned_sector, sector_per_transfer) maybe -1 if the sector to read is in one transfer block and not aligned to both side, but should still works
+			int32_t aligned_sector = count - unaligned_head_sector - unaligned_tail_sector;
+			sddf_count += DIV_POWER_OF_2(aligned_sector, sector_per_transfer);
+			
+			assert(MUL_POWER_OF_2(sddf_count, BLK_TRANSFER_SIZE) <= MAX_CLUSTER_SIZE);
+
+			#ifdef FS_DEBUG_PRINT
+			sddf_printf("blk_enqueue_read pre adjust: addr: 0x%lx sector: %u, count: %u ID: %d\n", read_data_offset, sector, count, handle);
+			sddf_printf("blk_enqueue_read after adjust: addr: 0x%lx sector: %u, count: %d ID: %d\n", read_data_offset, sddf_sector, sddf_count, handle);
+			#endif
+			blk_enqueue_req(blk_queue_handle, BLK_REQ_READ, read_data_offset, sddf_sector, sddf_count, handle);
```

Comment:
```suggestion
			int err = blk_enqueue_req(blk_queue_handle, BLK_REQ_READ, read_data_offset, sddf_sector, sddf_count, handle);
			assert(!err);
```

To confirm the invariant that the coroutine only ever enqueues a single block request.